ui/src/containers/BeamlineActionsContainer.jsx

```diff
@@ -122,98 +115,39 @@ class BeamlineActionsContainer extends React.Component {
-                ) : (
-                  <Button
-                    disabled={currentActionRunning}
-                    variant="primary"
-                    onClick={() => {
-                      this.startAction(currentActionName);
-                    }}
-                  >
-                    Run
-                  </Button>
-                )}
-              </Col>
-            </Row>
-            <hr />
-            <Plot1D
-              displayedPlotCallback={this.newPlotDisplayed}
-              plotId={this.plotIdByAction[currentActionName]}
-              autoNext={currentActionRunning}
-            />
-            {this.props.currentAction.messages.length > 0 ? (
-              <Card>
-                {this.props.currentAction.messages.map((message) => (
-                  // eslint-disable-next-line react/jsx-key
-                  <p>{message.message}</p>
-                ))}
-              </Card>
-            ) : (
-              ''
-            )}
-          </Modal.Body>
-          <Modal.Footer>
-            <Button
-              variant="outline-secondary"
-              onClick={this.hideOutput}
-              disabled={currentActionRunning}
-            >
-              Close window
-            </Button>
-          </Modal.Footer>
-        </DraggableModal>
+        {this.props.currentAction.argument_type === 'List' ? (
+          <BeamlineActionDialog
+            isDialogVisble={this.props.currentAction.show}
+            handleOnHide={this.hideOutput}
+            defaultPosition={defaultDialogPosition}
+            actionName={currentActionName}
+            actionArguments={this.props.currentAction.arguments}
+            isActionRunning={currentActionRunning}
+            actionMessages={this.props.currentAction.messages}
+            handleSetActionArgument={this.props.setArgumentValue}
+            handleStopAction={this.props.stopAction}
+            handleStartAction={this.startAction}
+            handleOnPlotDisplay={this.newPlotDisplayed}
+            plotId={this.plotIdByAction[currentActionName]}
+          />
+        ) : null}
+        {this.props.currentAction.argument_type === 'JSONSchema' ? (
+          <AnnotatedBeamlineActionDialog
+            isDialogVisble={this.props.currentAction.show}
+            handleOnHide={this.hideOutput}
+            defaultPosition={defaultDialogPosition}
+            actionName={this.props.currentAction.username}
+            actionId={currentActionName}
+            actionSchema={this.props.currentAction.schema}
+            isActionRunning={currentActionRunning}
+            actionMessages={this.props.currentAction.messages}
+            handleSetActionArgument={this.props.setArgumentValue}
+            handleStopAction={this.props.stopAction}
+            handleStartAction={this.props.startAction}
+            handleOnPlotDisplay={this.newPlotDisplayed}
+            plotId={this.plotIdByAction[currentActionName]}
+          />
+        ) : null}
```

Comment:
Seems like there's a lot of duplication between these two components. Perhaps a good case for composition with `children`?

Reply:
Hm, yeah why not. I find using `children` a bit unclear but thats perhaps the preferred way. I guess the are supposed to rendered in the order the are passed. Otherwise perhaps just creating some smaller components and reusing them in a more general dialog

Reply:
Yeah multiple ways to do this. The point is to abstract the beamline action dialog logic in a reusable component. How/where you use that dialog component is up to you. However, it's always good not to over optimise too soon, so I'd recommend starting by rendering that beamline action dialog once in `BeamlineActionContainer`, and rendering the two possible children inline via a ternary, also directly in `BeamlineActionContainer`.

Reply:
Sounds very reasonable, I finally did the later it seemed more intuitive.